### PR TITLE
add WebMapTileServiceRaster support (WMTS)

### PR DESCRIPTION
**Pull request by tal952**
Added WMTS support.
Based it on the WMS current support.

### CODE REVIEW

**Comment by tal952** [comment]
Hello @j9liu , Given that this Pull Request builds upon your prior contributions to the WMS support, could you kindly take a moment to review it? Your insights would be greatly appreciated.

**Comment by tal952** [comment]
He @kring , signed the Contributor License Agreement 👍

**Comment by kring** [comment]
Really appreciate the PR @tal952, and we'll review this as soon as we can. It might be a little bit though, we have a long review backlog at the moment.

File: Editor/CesiumWebMapTileServiceRasterOverlayEditor.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+                "The base URL of the Web Map Service (WMS)." +
+                "\n\n" +
+                "e.g." +
+                "\n\n" +
+                "https://services.ga.gov.au/gis/services/NM_Culture_and_Infrastructure/MapServer/WMSServer");
+            EditorGUILayout.DelayedTextField(this._baseUrl, baseUrlContent);
+
+            GUIContent tileWidthContent = new GUIContent(
+                "Tile Width",
+                "Image width.");
```
Comment (j9liu) [inline Editor/CesiumWebMapTileServiceRasterOverlayEditor.cs:77]:
```suggestion
                "The width of the image tiles in pixels.");
```

File: Editor/CesiumWebMapTileServiceRasterOverlayEditor.cs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            CesiumInspectorGUI.ClampedIntField(
+                this._tileWidth,
+                64,
+                2048,
+                tileWidthContent);
+
+            GUIContent tileHeightContent = new GUIContent(
+                "Tile Height",
+                "Image height.");
```
Comment (j9liu) [inline Editor/CesiumWebMapTileServiceRasterOverlayEditor.cs:86]:
```suggestion
                "The height of the image tiles in pixels.");
```

File: Editor/CesiumWebMapTileServiceRasterOverlayEditor.cs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+            EditorGUILayout.PropertyField(this._maximumLevel, maximumLevelContent);
+            
+            GUIContent formatContent = new GUIContent(
+                "Format",
+                "The MIME type for images to retrieve from the server, default value is \"image/jpeg\".");
+            EditorGUILayout.DelayedTextField(this._format, formatContent);
+            
+            GUIContent styleContent = new GUIContent(
+                "Style",
+                "The style name for WMTS requests, default value is \"default\".");
```
Comment (j9liu) [inline Editor/CesiumWebMapTileServiceRasterOverlayEditor.cs:114]:
```suggestion
                "The MIME type for images to retrieve from the server.");
            EditorGUILayout.DelayedTextField(this._format, formatContent);
            
            GUIContent styleContent = new GUIContent(
                "Style",
                "The style name for WMTS requests.");
```

File: Editor/CesiumWebMapTileServiceRasterOverlayEditor.cs [hunk 4]
```diff
@@ -0,0 +1,143 @@
+﻿using UnityEditor;
```
Comment (j9liu) [inline Editor/CesiumWebMapTileServiceRasterOverlayEditor.cs:1]: Could you reorder the properties so they match the order in Cesium for Unreal?

image

We like to have consistency between engines when possible, so people won't get confused between different UIs.

File: Runtime/CesiumWebMapTileServiceRasterOverlay.cs [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+            {
+                this._useGeographicProjection = value;
+                this.Refresh();
+            }
+        }
+
+        [SerializeField] private int _minimumLevel = 0;
+
+        /// <summary>
+        /// The minimum level-of-detail supported by the imagery provider, default value is 0.
```
Comment (j9liu) [inline Runtime/CesiumWebMapTileServiceRasterOverlay.cs:114]:
```suggestion
        /// The minimum level-of-detail supported by the imagery provider.
```

File: Runtime/CesiumWebMapTileServiceRasterOverlay.cs [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+            {
+                this._minimumLevel = value;
+                this.Refresh();
+            }
+        }
+
+        [SerializeField] private int _maximumLevel = 25;
+
+        /// <summary>
+        /// The maximum level-of-detail supported by the imagery provider, default value is 25.
```
Comment (j9liu) [inline Runtime/CesiumWebMapTileServiceRasterOverlay.cs:129]:
```suggestion
        /// The maximum level-of-detail supported by the imagery provider.
```

**Review by j9liu — CHANGES_REQUESTED** [review]
Thanks @tal925 for the PR! This is a solid start, I reviewed and left some feedback where needed. Mainly, I'm concerned about the difference in overlay parameters between this PR and Cesium for Unreal. Other comments are just minor style changes / nitpicks.

Do you have a test URL that you're willing to share, so that I can verify it works?

File: Runtime/CesiumWebMapTileServiceRasterOverlay.cs [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+                this.Refresh();
+            }
+        }
+        
+        [SerializeField] private bool _useGeographicProjection = false;
+
+        /// <summary>
+        /// Change projection to GeographicProjection (the default is WebMercatorProjection).
+        /// </summary>
+        public bool useGeographicProjection
```
[the remark this answers is not in the document]
Comment (tal952) [reply Runtime/CesiumWebMapTileServiceRasterOverlay.cs:101]: That is very strange, the default in CesiumJS and Cesium-native is the WebMercatorProjection, I'm not sure why they made useWebMercatorProjection and not useGeographicProjection.
It looks like we need to decide between CesiumJS and Cesium-native default and Cesium-Unreal.
I think we should go with the CesiumJS and Cesium-native, what do you think?

File: Runtime/CesiumWebMapTileServiceRasterOverlay.cs [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+                this.Refresh();
+            }
+        }
+
+        [SerializeField] private string _tileMatrixSetID = "";
+
+        /// <summary>
+        /// The identifier of the TileMatrixSet to use for WMTS requests.
+        /// </summary>
+        public string tileMatrixSetID
```
[the remark this answers is not in the document]
Comment (tal952) [reply Runtime/CesiumWebMapTileServiceRasterOverlay.cs:86]: It's very uncommon to use this fields, I think that if anyone needs them, they just implement them. (It's just kind of lot of work to add this uncommon fields)

File: native~/Runtime/src/CesiumWebMapTileServiceRasterOverlayImpl.cpp [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  wmtsOptions.style = overlay.style().ToStlString();
+  wmtsOptions.tileMatrixSetID = overlay.tileMatrixSetID().ToStlString();
+  wmtsOptions.minimumLevel = overlay.minimumLevel();
+  wmtsOptions.maximumLevel = overlay.maximumLevel();
+  wmtsOptions.tileWidth = overlay.tileWidth();
+  wmtsOptions.tileHeight = overlay.tileHeight();
+
+  if (overlay.useGeographicProjection()) {
+    wmtsOptions.projection = CesiumGeospatial::GeographicProjection();
+  }
```
[the remark this answers is not in the document]
Comment (tal952) [reply native~/Runtime/src/CesiumWebMapTileServiceRasterOverlayImpl.cpp:58]: https://github.com/CesiumGS/cesium-unity/pull/430#discussion_r1536776088

File: Runtime/CesiumWebMapTileServiceRasterOverlay.cs [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    public partial class CesiumWebMapTileServiceRasterOverlay : CesiumRasterOverlay
+    {
+        [SerializeField] private string _baseUrl = "";
+
+        /// <summary>
+        /// The base URL of the Web Map Tile Service (WMTS).
+        /// </summary>
+        /// <remarks>
+        /// This URL should not include query parameters. For example:
+        /// https://services.ga.gov.au/gis/services/NM_Culture_and_Infrastructure/MapServer/WMTSServer
```
[the remark this answers is not in the document]
Comment (tal952) [reply Runtime/CesiumWebMapTileServiceRasterOverlay.cs:24]: Changed to https://tile.openstreetmap.org/{TileMatrix}/{TileCol}/{TileRow}.png

File: Runtime/CesiumWebMapTileServiceRasterOverlay.cs [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+        /// </summary>
+        public string style
+        {
+            get => this._style;
+            set
+            {
+                this._style = value;
+                this.Refresh();
+            }
+        }
```
[the remark this answers is not in the document]
Comment (tal952) [reply Runtime/CesiumWebMapTileServiceRasterOverlay.cs:64]: AFAIK it's very standard to have "default" style, but I'm not sure about that. I'll leave it empty.

File: Runtime/CesiumWebMapTileServiceRasterOverlay.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (j9liu) [reply Runtime/CesiumWebMapTileServiceRasterOverlay.cs:101]: Oh that's strange... Let's keep the settings consistent with CesiumJS and Cesium Native. Thank you for pointing that out to me!

File: Runtime/CesiumWebMapTileServiceRasterOverlay.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (j9liu) [reply Runtime/CesiumWebMapTileServiceRasterOverlay.cs:86]: Sounds good @tal952. Thanks for letting me know 👍

**Comment by j9liu** [comment]
@tal952 thanks for the responses to my comments above. Let me know when I should take another look!

**Comment by tal952** [comment]
Hey @j9liu , looks like I fixed everything you mentioned, if you can take another look it would be awesome 🙏🙏

**Comment by j9liu** [comment]
Looks good to me @tal952, thanks for the contribution! I'll do a follow-up PR to add those extra settings to the overlay, just for parity with Unreal.

**Pull request merged**